### ISSUE TITLE
RFC: The `Standard Schema` spec

### ISSUE THREAD

**Issue by colinhacks** [standard-schema/standard-schema#1]
Hello all! I wrote up this proposal and would love to get some early feedback from my fellow validation library authors. 🤙

cc @sinclairzx81 @ssalbdivad @fabian-hiller @gcanti @jquense @ianstormtaylor @davidmdm

**Comment by fabian-hiller** [comment]
Thank you Colin for this contribution to the community! I am currently very busy rewriting Valibot. As soon as I find some time, I will give you feedback on everything. For example, I don't think the current definition of `path` is sufficient. For records and maps it is necessary to know if a path element points to the key or the value. That's why Valibot uses an object with additional info for each path item.

```ts
interface PathItem {
  schema: string;
  origin: 'key' | 'value'; // <--
  input: unknown;
  key: string | number | symbol | undefined;
  value: unknown;
}
```

**Comment by fabian-hiller** [comment]
@decs this might be of interest to you.

**Comment by decs** [comment]
Thanks for kickstarting this effort, Colin! We're on the same page that this is something worth solving.

One challenge we had on [typeschema](https://github.com/decs/typeschema) was that many validation libraries have schemas without a validate function (i.e. `schema.parse(...)` doesn't exist). On Valibot, for example, it has to be `parse(schema, ...)`. We solved it by [wrapping that logic](https://github.com/decs/typeschema/blob/main/packages/valibot/src/validation.ts#L16) into another function. A possible solution could be to ask library maintainers who took that route to provide a similar wrapping function that outputs a `StandardSchema`.

It's also worth considering how async validation should be handled. Should `~validate` allow returning a Promise?

Have you considered how `StandardSchema` can evolve over time? Validation is the base case, but some form libraries also rely on introspection (decs/typeschema#9). So I expect that would be the next feature request.

I ended up writing a [very similar spec](https://github.com/decs/typeschema/blob/main/packages/core/src/wrap.ts#L5-L13) for interfacing with tRPC. I'm open to replacing that with `StandardSchema` (once we reach a more stable spec) to help on this migration period. That way, typeschema's `wrap(schema)` can be an alternative for validation libraries that haven't implemented `StandardSchema` yet.

Looking forward to making this a reality!

**Comment by ssalbdivad** [comment]
@fabian-hiller A standard like this needs to require only the most essential parts of the API that any validation library would almost have to implement in some form by definition.

There is certainly room to take things further e.g. in terms of error context or introspection as mentioned by @decs, but if they are to be added to the spec, they should be optional so library consumers can appropriately handle those methods/properties being unimplemented.

Potentially, the spec could also provide a type with those enhancements required, and consumers requiring capabilities like introspection could accept a type like `IntrospectableSchema` from `standard-schema` which would extend the minimal validation API.

**Comment by colinhacks** [comment]
> For example, I don't think the current definition of path is sufficient. For objects, records and maps it is necessary to know if a path item points to the key or the value.

Good point. Zod [just adds a literal `"key"` or `"value"`](https://github.com/colinhacks/zod/blob/78e5cde0d7b3ce4b75dc3395cf86c2ca0af0d167/src/types.ts#L3660-L3666) to the path for ZodMap errors but I certainly see the limits of that. That said, I think a special `origin` key that only applies to a small subset of schema types isn't ideal. As David says, I think this case may not merit special treatment. You can always clarify the nature of the issue in the `message`.

> @decs this might be of interest to you and [typeschema](https://github.com/decs/typeschema).

Ah of course! Hi @decs!

> One challenge we had on [typeschema](https://github.com/decs/typeschema) was that many validation libraries have schemas without a validate function (i.e. schema.parse(...) doesn't exist). On Valibot, for example, it has to be parse(schema, ...)

Hm interesting. In the case of valibot this doesn't seem to be a problem. Schemas already have an internal `_parse` method as in Zod. Plus parsing logic is self-contained within each schema type, so I don't think there are treeshakingn considerations. Obviously @fabian-hiller should chime in if I'm wrong about this.

**Comment by colinhacks** [comment]
> It's also worth considering how async validation should be handled

Great question. My first thought is that `~validate()` should probably return `T | Promise<T>`. The validation library determines which one it'll be conditional on whether the schema contains any async validation logic, and the consuming library (tRPC, etc) just has to handle both cases. If the consuming library requires a synchronous validator, they can throw an informative runtime error should the schema return a `Promise`.

> Validation is the base case, but some form libraries also rely on introspection 

We certainly don't want to design a new serializable schema specification. That way lies madness. I think it would be a *huge* win if all libraries implemented a standard method for JSON Schema serialization.

```ts
interface StandardSchema {
  toJsonSchema(): JsonSchema;
}
```

But that's a big lift for library authors. There are a lot of implementation-specific nuances w.r.t how to do the conversion, e.g. should an `email` field be represented with `format:email`, `format:idn-email`, or a custom regex? See the Options section for [zod-to-json-schema](https://github.com/StefanTerdell/zod-to-json-schema?tab=readme-ov-file#options) for a sense of the complexity there. 

We would also hit against the "no method" problem here. I imagine most schema authors would be resistant to encapsulating `toJsonSchema` logic inside their schema types for tree shaking reasons.

**Comment by colinhacks** [comment]
> There is certainly room to take things further e.g. in terms of error context or introspection as mentioned by @decs, but if they are to be added to the spec, they should be optional so library consumers can appropriately handle those methods/properties being unimplemented.
>
>Potentially, the spec could also provide a type with those enhancements required, and consumers requiring capabilities like introspection could accept a type like IntrospectableSchema from standard-schema which would extend the minimal validation API.

+1 to all of this

**Comment by fabian-hiller** [comment]
> That said, I think a special `origin` key that only applies to a small subset of schema types isn't ideal. As David says, I think this case may not merit special treatment. You can always clarify the nature of the issue in the `message`.

I agree that this probably only applies to a small subset of schemas. Therefore, I understand making a tradeoff here. However, for multiple nested maps or other similar structures, you need to know the origin of each path item to be able to navigate through the datastructure. I just question whether a standard should be limited and in some cases misleading in this regard.

> Hm interesting. In the case of valibot this doesn't seem to be a problem. Schemas already have an internal `_parse` method as in Zod.

I think there might be some problems when Valibots internal `_parse` (or in the future `_run`) method needs to apply to `'~validate': (data: unknown) => T | ValidationError;` because Valibot also needs to track if a value is typed. That's why Valibots use a return type similar to this `{ typed: boolean; value: Value | unknown; issues?: [Issue, ...Issue[]] }`. Adding an additional method just to apply to to spec could be a problem in terms of bundle size.

But as @decs mentioned, a solution for Valibot might be to provide an official adpater that applies to this spec.

**Comment by StefanTerdell** [comment]
> That way lies madness

You called? 😄
